File: kiko-api/src/services/swap/handleSwapError.ts

import { logger } from '../../utils/logger.js';
import { LogCode } from '../../config/logRegistry.js';

/**
 * Handle and map swap errors to user-friendly messages.
 * Maps technical errors from Jupiter and 0x to professional English phrases.
 */
export function handleSwapError(error: any): string {
    const msg = error.message || 'Unknown error';
    const msgLower = msg.toLowerCase();
    let errorData: any = error.response?.data || error.data || {};

    if (
        msgLower.startsWith('copytrade_fallback_guard_reject:')
        || msgLower.startsWith('quote_anchor_guard_reject:')
    ) {
        return msg;
    }

    // Extract JSON payload from fetchJson errors (format: "HTTP 400: {...}")
    if (!errorData || Object.keys(errorData).length === 0) {
        const httpJsonMatch = msg.match(/HTTP\s+\d+:\s*(\{.*\})/);
        if (httpJsonMatch) {
            try {
                errorData = JSON.parse(httpJsonMatch[1]);
            } catch {
                errorData = {};
            }
        }
    }

    // Log the raw technical error for debugging purposes
    logger.error(LogCode.EXE_TX_REVERTED, 'Raw Swap Error Captured', { msg, errorData });

    // 1. Slippage & Price Movement Errors
    if (
        msgLower.includes('slippage') ||
        msgLower.includes('price_or_slippage_too_low') ||
        msgLower.includes('insufficient_output_amount') ||
        msgLower.includes('uniswapv2: k') ||
        msgLower.includes('too little received') ||
        msgLower.includes('minamountout') ||
        msgLower.includes('min output') ||
        msgLower.includes('err_limit_out') ||
        msgLower.includes('0x1771') // Jupiter slippage error code
    ) {
        return 'Swap failed: Price moved beyond slippage tolerance. Please try increasing your slippage percentage.';
    }

    // 2. Liquidity & Market Depth Errors
    if (
        msgLower.includes('insufficient_asset_liquidity') ||
        msgLower.includes('insufficient_liquidity') ||
        msgLower.includes('no route found') ||
        msgLower.includes('no route matched') ||
        msgLower.includes('route_not_found') ||
        msgLower.includes('pool_not_found') ||
        msgLower.includes('pair_not_found') ||
        msgLower.includes('could_not_fill')
    ) {
        return 'Swap failed: Insufficient market liquidity or depth for this trade scale. Try reducing the amount.';
    }

    // 3. Balance & Gas Fee Errors
    if (
        msgLower.includes('insufficient funds') ||
        msgLower.includes('insufficient balance') ||
        msgLower.includes('insufficient funds for gas') ||
        msgLower.includes('insufficient funds for intrinsic transaction cost') ||
        msgLower.includes('erc20: transfer amount exceeds balance') ||
        msgLower.includes('erc20insufficientbalance') ||
        msgLower.includes('insufficient lamports') ||
        msgLower.includes('balance')
    ) {
        return 'Swap failed: Insufficient balance for the transaction or required gas fees. Ensure you have enough native tokens.';
    }

    // 4. Permission / Allowance Errors (EVM)
    if (
        msgLower.includes('approved spender no longer matches') ||
        msgLower.includes('quote changed after approval') ||
        msgLower.includes('explicit approval quote')
    ) {
        return 'Swap failed: The route changed after token approval, so no swap transaction was sent. Please retry to get a fresh quote.';
    }

    if (
        msgLower.includes('allowance_too_low') ||
        msgLower.includes('insufficient allowance') ||
        (msgLower.includes('execution reverted') && msgLower.includes('allowance')) ||
        msgLower.includes('transfer_from_failed') ||
        msgLower.includes('transferhelper')
    ) {
        return 'Swap failed: Token approval or allowance is insufficient. Please approve the token and retry.';
    }

    // 4.5 Amount too small / invalid
    if (
        msgLower.includes('amount too low') ||
        msgLower.includes('amount is too low') ||
        msgLower.includes('invalid amount') ||
        msgLower.includes('sellamount too low') ||
        msgLower.includes('min buy amount') ||
        msgLower.includes('min buy amount not met')
    ) {
        return 'Swap failed: Amount too small for this pair or below minimum. Try a larger amount.';
    }

    // 5. User-Initiated Cancellation
    if (msgLower.includes('user rejected') || msgLower.includes('user denied') || msgLower.includes('rejected')) {
        return 'Swap failed: Transaction was rejected by the user.';
    }

    // 6. Network Connectivity & Timeouts
    if (msgLower.includes('timeout') || msgLower.includes('etimedout')) {
        return 'Swap failed: Network request timed out. Please check your connection and try again later.';
    }

    // Generic fallback for unhandled technical errors
    return `Swap failed: ${msg.substring(0, 100)}`;
}

File: kikoapi/internal/services/swap/handle_swap_error.go
package swap

import (
	"regexp"
	"strings"
)

// HandleSwapError maps technical errors from Jupiter, 0x, Kyber to user-friendly messages.
func HandleSwapError(err interface{}) string {
	var msg string
	switch v := err.(type) {
	case error:
		msg = v.Error()
	case string:
		msg = v
	default:
		msg = "Unknown error"
	}
	msgLower := strings.ToLower(msg)

	// Kyber codes (would come from errorData in TS - here we only have message)
	if strings.Contains(msgLower, "unable to bind request body") {
		return "Swap failed: Kyber request payload invalid. Retrying may help, but this is likely a routing/build error."
	}
	if strings.Contains(msgLower, "route not found") || strings.Contains(msgLower, "route_not_found") {
		return "Swap failed: No Kyber route found for this pair. Try a smaller amount or different pair."
	}
	if strings.Contains(msgLower, "amount exceeds") && strings.Contains(msgLower, "maximum") {
		return "Swap failed: Amount exceeds maximum for this route. Try a smaller amount."
	}
	if strings.Contains(msgLower, "no eligible pool found") {
		return "Swap failed: No eligible pool found for this route. Try a different pair or smaller amount."
	}
	if strings.Contains(msgLower, "tokenin or tokenout not supported") {
		return "Swap failed: tokenIn or tokenOut not supported by Kyber."
	}

	// Slippage & price movement
	if strings.Contains(msgLower, "slippage") ||
		strings.Contains(msgLower, "price_or_slippage_too_low") ||
		strings.Contains(msgLower, "insufficient_output_amount") ||
		strings.Contains(msgLower, "uniswapv2: k") ||
		strings.Contains(msgLower, "too little received") ||
		strings.Contains(msgLower, "minamountout") ||
		strings.Contains(msgLower, "min output") ||
		strings.Contains(msgLower, "err_limit_out") ||
		strings.Contains(msgLower, "0x1771") {
		return "Swap failed: Price moved beyond slippage tolerance. Please try increasing your slippage percentage."
	}

	// Liquidity & market depth
	if strings.Contains(msgLower, "insufficient_asset_liquidity") ||
		strings.Contains(msgLower, "insufficient_liquidity") ||
		strings.Contains(msgLower, "no route found") ||
		strings.Contains(msgLower, "no route matched") ||
		strings.Contains(msgLower, "route_not_found") ||
		strings.Contains(msgLower, "pool_not_found") ||
		strings.Contains(msgLower, "pair_not_found") ||
		strings.Contains(msgLower, "could_not_fill") {
		return "Swap failed: Insufficient market liquidity or depth for this trade scale. Try reducing the amount."
	}

	// Balance & gas
	if strings.Contains(msgLower, "insufficient funds") ||
		strings.Contains(msgLower, "insufficient balance") ||
		strings.Contains(msgLower, "insufficient funds for gas") ||
		strings.Contains(msgLower, "insufficient funds for intrinsic transaction cost") ||
		strings.Contains(msgLower, "erc20: transfer amount exceeds balance") ||
		strings.Contains(msgLower, "erc20insufficientbalance") ||
		strings.Contains(msgLower, "insufficient lamports") ||
		strings.Contains(msgLower, "balance") {
		return "Swap failed: Insufficient balance for the transaction or required gas fees. Ensure you have enough native tokens."
	}

	// Allowance
	if strings.Contains(msgLower, "allowance_too_low") ||
		strings.Contains(msgLower, "insufficient allowance") ||
		(strings.Contains(msgLower, "execution reverted") && strings.Contains(msgLower, "allowance")) ||
		strings.Contains(msgLower, "transfer_from_failed") ||
		strings.Contains(msgLower, "transferhelper") {
		return "Swap failed: Token approval or allowance is insufficient. Please approve the token and retry."
	}

	// Amount too small
	if strings.Contains(msgLower, "amount too low") ||
		strings.Contains(msgLower, "amount is too low") ||
		strings.Contains(msgLower, "invalid amount") ||
		strings.Contains(msgLower, "sellamount too low") ||
		strings.Contains(msgLower, "min buy amount") ||
		strings.Contains(msgLower, "min buy amount not met") {
		return "Swap failed: Amount too small for this pair or below minimum. Try a larger amount."
	}

	// User rejected
	if strings.Contains(msgLower, "user rejected") || strings.Contains(msgLower, "user denied") || strings.Contains(msgLower, "rejected") {
		return "Swap failed: Transaction was rejected by the user."
	}

	// Timeout
	if strings.Contains(msgLower, "timeout") || strings.Contains(msgLower, "etimedout") {
		return "Swap failed: Network request timed out. Please check your connection and try again later."
	}

	// Extract JSON from "HTTP 400: {...}" for future use
	_ = regexp.MustCompile(`HTTP\s+\d+:\s*(\{.*\})`)

	if len(msg) > 100 {
		msg = msg[:100]
	}
	return "Swap failed: " + msg
}
